apps/src/gamelab/GameLab.js

```diff
@@ -826,38 +826,132 @@ GameLab.prototype.onP5ExecutionStarting = function () {
-
-  this.initInterpreter();
-  // And execute the interpreter for the first time:
-  if (this.JSInterpreter && this.JSInterpreter.initialized()) {
-    // Start executing the interpreter's global code as long as a setup() method
-    // was provided. If not, we will skip running any interpreted code in the
-    // preload phase and wait until the setup phase.
-    this.preloadInProgress = true;
-    if (this.globalCodeRunsDuringPreload) {
-      this.JSInterpreter.executeInterpreter(true);
-      this.interpreterStarted = true;
+  let isReadyNow = false;
+  Promise.all([
+      this.preloadAnimations_(),
+      this.runPreloadEventHandler_()
+  ]).then(() => {
+    isReadyNow = true;
+    this.gameLabP5.notifyPreloadPhaseComplete();
+  });
+  return isReadyNow;
+};
 
-      // In addition, execute the global function called preload()
-      if (this.eventHandlers.preload) {
-        this.eventHandlers.preload.apply(null);
-      }
+/**
+ * Wait for animations to be loaded into memory and ready to use, then pass
+ * those animations to P5 to be loaded into the engine as animations.
+ * @returns {Promise} which resolves once animations are in memory in the redux
+ *          store and we've started loading them into P5.
+ *          Loading to P5 is also an async process but it has its own internal
+ *          effect on the P5 preloadCount, so we don't need to track it here.
+ * @private
+ */
+GameLab.prototype.preloadAnimations_ = function () {
+  let store = this.studioApp_.reduxStore;
+  return new Promise(resolve => {
+    if (this.areAnimationsReady_()) {
+      resolve();
     } else {
-      if (this.eventHandlers.preload) {
-        this.log("WARNING: preload() was ignored because setup() was not provided");
-        this.eventHandlers.preload = null;
+      // Watch store changes until all the animations are ready.
+      const unsubscribe = store.subscribe(() => {
+        if (this.areAnimationsReady_()) {
+          unsubscribe();
+          resolve();
+        }
+      });
+    }
+  }).then(() => {
+    // Animations are ready - send them to p5 to be loaded into the engine.
+    this.gameLabP5.preloadAnimations(store.getState().animationList);
+  });
+};
+
+/**
+ * Check whether all animations in the project animation list have been loaded
+ * into memory and are ready to use.
+ * @returns {boolean}
+ * @private
+ */
+GameLab.prototype.areAnimationsReady_ = function () {
+  const animationList = this.studioApp_.reduxStore.getState().animationList;
+  return animationList.orderedKeys.every(key => animationList.propsByKey[key].loadedFromSource);
+};
+
+/**
+ * Run the preload event handler, and optionally global code, and report when
+ * it is done by resolving a returned Promise.
+ * @returns {Promise} Which will resolve immediately if there is no code to run,
+ *          otherwise will resolve when the preload handler has completed.
+ * @private
+ */
+GameLab.prototype.runPreloadEventHandler_ = function () {
```

Comment:
So, gamelab (and p5play) has several kinds of magic handlers.  The one we see most often (and the only one we expect students to use is `draw()` but it's not the only one - you've also go `preload()` and `setup()`, important lifecycle methods in getting everything up and running.

In reality, when you work with p5play you have write your code like this:

```javascript
// Global setup work, doesn't know anything about p5

function preload() {
  // Can do some initial p5 setup (like loading assets) but lots of stuff not available yet
}

function setup() {
  // Setup work, which has access to the p5 world.
}

function draw() {
  // run loop
}
```

But we've introduced some convenience modes in gamelab that p5play doesn't have (and I think we also provide hooks in levelbuilder to add hidden code to preload and setup).  At first we had one where anything in the 'global' scope was run inside preload:

```javascript
// This stuff is implicitly in p5's preload()

function setup() {
  // Setup work, which has access to the p5 world.
}

function draw() {
  // run loop
}
```

That got further revised and we added the current default mode, where stuff in global scope becomes the setup method - because it's easy to work with:

```javascript
 // This stuff is implicitly run in p5's setup()

function draw() {
  // run loop
}
```

Even in this mode, we use `preload()` under the hood to get animations loaded before any student code gets run, and as a levelbuilder hook.